### PR TITLE
Implement AMP version 2

### PR DESCRIPTION
**Pull request by jothan**
Implement AMP v2 as described here: https://amp-protocol.net/conversations_v2.html

## Contributor Checklist:

* [x] The associated tickets in Trac are here:
 * https://twistedmatrix.com/trac/ticket/2529
 * https://twistedmatrix.com/trac/ticket/9488
* [x] I ran `tox -e black-reformat` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

File: src/twisted/protocols/amp.py [hunk 1]
```diff
@@ -657,7 +658,7 @@ def copy(self):
         newBox.update(self)
         return newBox
 
-    def serialize(self):
+    def serialize(self, version2=False):
```
Comment (rodrigc) [inline src/twisted/protocols/amp.py]: Can you add type annotations for any new functions which you have added or modified?  The Twisted tree is slowly moving to add type annotations

File: src/twisted/protocols/amp.py [hunk 2]
```diff
@@ -2613,6 +2634,35 @@ def connectionLost(self, reason):
         self.transport = None
 
 
+class AMPv2(AMP):
+    def _write_box(self, box):
```
Comment (rodrigc) [inline src/twisted/protocols/amp.py]: Annotate?

File: src/twisted/protocols/amp.py [hunk 3]
```diff
@@ -2613,6 +2634,35 @@ def connectionLost(self, reason):
...
+        self.transport.write(box.serialize(version2=True))
+
+    def proto_value(self, string):
```
Comment (rodrigc) [inline src/twisted/protocols/amp.py]: Annotate?

File: src/twisted/protocols/amp.py [hunk 4]
```diff
@@ -2613,6 +2634,35 @@ def connectionLost(self, reason):
...
+        if len(string) < MAX_VALUE_LENGTH:
+            self._currentBox[self._currentKey] = string
+            self._currentKey = None
+            self.MAX_LENGTH = self._MAX_KEY_LENGTH
+            return "key"
+
+        self._currentValue = [string]
+        return "valuecont"
+
+    def proto_valuecont(self, string):
```
Comment (rodrigc) [inline src/twisted/protocols/amp.py]: Annotate?

File: src/twisted/test/test_amp.py [hunk 5]
```diff
@@ -95,6 +95,14 @@ def amp_HELLO(self, box):
         return amp.Box(hello=box[b"hello"])
 
 
+class SimpleSymmetricProtocol2(amp.AMPv2):
+    def sendHello(self, text):
```
Comment (rodrigc) [inline src/twisted/test/test_amp.py]: Annotate?

File: src/twisted/test/test_amp.py [hunk 6]
```diff
@@ -95,6 +95,14 @@ def amp_HELLO(self, box):
...
+        return self.callRemoteString(b"hello", hello=text)
+
+    def amp_HELLO(self, box):
```
Comment (rodrigc) [inline src/twisted/test/test_amp.py]: Annotate?

**Review by rodrigc — CHANGES_REQUESTED** [review]
Thanks for the contribution

**Comment by jothan** [comment]
@rodrigc I have added the type annotations as you requested. May I resolve the comments you added ?

File: src/twisted/protocols/amp.py [hunk 7]
```diff
@@ -2613,6 +2634,35 @@ def connectionLost(self, reason):
         self.transport = None
 
 
+class AMPv2(AMP):
```
Comment (glyph) [inline src/twisted/protocols/amp.py:2646]: This class and all added methods here need docstrings.

**Comment by rodrigc** [comment]
@jothan At the top-level Twisted directory, you can run `tox -e mypy` to check things before mypy before pushing to GitHub.  That will save a few cycles.

**Comment by jothan** [comment]
@rodrigc Thanks for the tip. However, I'm a bit stumped by the remaining errors, do you have any tips ?

File: src/twisted/protocols/amp.py [hunk 8]
```diff
@@ -657,15 +658,15 @@ def copy(self):
...
+    def serialize(self, version2: bool = False) -> bytes:
         """
         Convert me into a wire-encoded string.
 
         @return: a C{bytes} encoded according to the rules described in the
             module docstring.
         """
         i = sorted(self.items())
-        L = []
+        L: List[bytes] = []
```
Comment (rodrigc) [inline src/twisted/protocols/amp.py]: This needs to be:

```
L = []  # type: List[bytes]
```

because Twisted still supports Python 3.5 which does not support the new syntax for type annotations on variables.  Hopefully by the end of the year, Twisted will be able to drop Python 3.5 support.

**Comment by jothan** [comment]
Alright, I figured out which type annotations I needed, mypy is now happy.

**Review by rodrigc — CHANGES_REQUESTED** [review]

**Comment by rodrigc** [comment]
Thanks for being patient and submitting the annotations.  This is slowly making the Twisted codebase better, and easier to understand.

File: src/twisted/protocols/amp.py (hunk 8 above)
Comment (jothan) [reply src/twisted/protocols/amp.py]: Ok for this change, but how do I fix this for Python 3.5 ?

```python
class AmpBox(Dict[bytes, bytes]):
```

**Comment by jothan** [comment]
@rodrigc I totally understand the need for those annotations, especially in a large code base.

My AMP implementation is in Rust with types everywhere ;P https://github.com/jothan/amp-async

File: src/twisted/protocols/amp.py (hunk 8 above)
Comment (jothan) [reply src/twisted/protocols/amp.py]: Oh, Python 3.5 seems perfectly happy with that class declaration.

**Review by rodrigc — APPROVED** [review]
I think this looks OK, but I'll leave open for another reviewer to give it a once over.

Nice to see you have experience writing AMP protocol parsers in Rust.  I need to learn Rust one of these days. :)

File: src/twisted/newsfragments/9488.feature [hunk 9]
```diff
@@ -0,0 +1 @@
+Implement the AMP version 2 protocol to allow values greater than 64k.
```
Comment (glyph) [inline src/twisted/newsfragments/9488.feature:1]: We shouldn't have duplicate NEWS entries like this, as it produces useless duplicative noise in the changelog.  The correct thing to do here is to close the second ticket as a duplicate in trac, then have there be one NEWS file.

File: src/twisted/protocols/amp.py [hunk 10]
```diff
@@ -239,6 +239,7 @@ def trapZero(result):
 
 __all__ = [
     "AMP",
+    "AMPv2",
```
Comment (glyph) [inline src/twisted/protocols/amp.py:242]: This should not be exposed as a separate class, but rather a protocol-version parameter.

File: src/twisted/protocols/amp.py [hunk 11]
```diff
@@ -608,7 +609,7 @@ class IncompatibleVersions(AmpError):
 PROTOCOL_ERRORS = {UNHANDLED_ERROR_CODE: UnhandledCommand}
 
 
-class AmpBox(dict):
+class AmpBox(Dict[bytes, bytes]):
```
Comment (glyph) [inline src/twisted/protocols/amp.py:612]: We really should have landed type annotations as a separate, prior change. They're not related to "AMPv2".

File: src/twisted/protocols/amp.py [hunk 12]
```diff
@@ -657,15 +658,15 @@ def copy(self):
         newBox.update(self)
         return newBox
 
-    def serialize(self):
+    def serialize(self, version2: bool = False) -> bytes:
```
Comment (glyph) [inline src/twisted/protocols/amp.py:661]: This "version2" parameter needs a `@param` in the docstring.

File: src/twisted/protocols/amp.py [hunk 13]
```diff
@@ -674,11 +675,29 @@ def serialize(self):
...
-                w(kv)
+
+            w(pack("!H", len(k)))
+            w(k)
+
+            if version2:
+                vio = BytesIO(v)
+                del v
+
+                # If the value is an exact multiple of 65535, the last
```
Comment (glyph) [inline src/twisted/protocols/amp.py:688]: how is the value "an exact multiple" of 65535 if it is not 65535 itself?

File: src/twisted/protocols/amp.py [hunk 14]
```diff
@@ -674,11 +675,29 @@ def serialize(self):
...
+                # chunk containing data will be followed by a zero-length chunk
+                # (i.e. when read() returns EOF).
+                chunk = vio.read(MAX_VALUE_LENGTH)
+                while True:
+                    w(pack("!H", len(chunk)))
```
Comment (glyph) [inline src/twisted/protocols/amp.py:693]: I realize that this is coming from the "spec" on amp-protocol.net, but there were never any discussions I saw about how to implement a V2.  Given that this is neither backwards compatible nor has any wire-level negotiation it seems like a pretty bad feature for a version bump.

Comment (glyph) [reply src/twisted/protocols/amp.py:693]: (A better way to have done this would be to normalize the meaning of repeated keys, but I digress.)

File: src/twisted/protocols/amp.py [hunk 15]
```diff
@@ -868,13 +889,17 @@ def _sendBoxCommand(self, command, box, requiresAnswer=True):
...
             result = self._outstandingRequests[tag] = Deferred()
-        else:
-            result = None
+
         return result
 
-    def callRemoteString(self, command, requiresAnswer=True, **kw):
+    def callRemoteString(
+        self, command: bytes, requiresAnswer: bool = True, **kw: bytes
+    ) -> Optional[Deferred]:
```
Comment (glyph) [inline src/twisted/protocols/amp.py:902]: `Optional[Deferred]` is the wrong way to specify this signature; what you want is an `@overload` to correlate the value of `requiresAnswer` to the type of the response.

File: src/twisted/protocols/amp.py [hunk 16]
```diff
@@ -2613,6 +2643,52 @@ def connectionLost(self, reason):
...
+    """
+
+    _currentValue = None  # type: Optional[List[bytes]]
+
+    def _write_box(self, box: AmpBox) -> None:
+        """Send an AmpBox using the AMPv2 format."""
+        assert self.transport is not None
+        self.transport.write(box.serialize(version2=True))
+
+    def proto_value(self, string: bytes) -> str:
```
Comment (glyph) [inline src/twisted/protocols/amp.py:2658]: These methods do not belong on an `AMP` subclass, if they belong on a subclass at all, but a `BinaryBoxProtocol` subclass.

There's a mea-culpa here too; the inheritance-for-convenience design of AMP was a mistake, and should not have created this dichotomy; but, this is also why this should not be a new class, though, and instead just a parameter to `BinaryBoxProtocol` that `AMP` can propagate.

**Review by glyph — COMMENTED** [review]
I'm sorry that I didn't have time to get in a code review before this landed, but my attention was brought to it and I have some post-hoc feedback.

File: src/twisted/newsfragments/9488.feature (hunk 9 above)
Comment (jothan) [reply src/twisted/newsfragments/9488.feature:1]: You might want to update the guidelines in this case, from the news fragment instructions: " You should replace <ticket number> with the ticket number which is being resolved by the change (if multiple tickets are resolved, multiple files with the same contents should be added). "

File: src/twisted/protocols/amp.py (hunk 11 above)
Comment (jothan) [reply src/twisted/protocols/amp.py:612]: I know, sorry for the noise, I wanted to follow trunk closely with tests passing.

File: src/twisted/protocols/amp.py (hunk 14 above)
Comment (jothan) [reply src/twisted/protocols/amp.py:693]: Well, it is _mostly_ backwards compatible as long as values are smaller than 65534 bytes. Considering it is a breaking change, it should be signaled via a compatibility version change. I would have called AMP v1.1, but AMP v2 already existed conceptually.

File: src/twisted/protocols/amp.py (hunk 13 above)
Comment (jothan) [reply src/twisted/protocols/amp.py:688]: Guess I should remove the word "exact" here.

File: src/twisted/protocols/amp.py (hunk 16 above)
Comment (jothan) [reply src/twisted/protocols/amp.py:2658]: Yeah, I got hit by that. My initial unpublished patch was version parameters everywhere.

**Comment by glyph** [comment]
@jothan in case you are not on the Twisted mailing list, I began a discussion about this here: https://twistedmatrix.com/pipermail/twisted-python/2020-October/065288.html

I'm leaning towards doing a revert based on that.

Thanks for your contribution to Twisted, sorry to burden it with more work!

File: src/twisted/newsfragments/9488.feature (hunk 9 above)
Comment (glyph) [reply src/twisted/newsfragments/9488.feature:1]: @jothan Hmm. Apparently I'm wrong about this.  I thought that this was an idiom that predated our usage of towncrier, but apparently towncrier adopted this same idiom of hashing the contents of the files; I just created a sample project to test it out and it does the right thing in all the edge cases that I thought would not be handled.  Sorry for the mistake!  I am still probably going to be rolling this back out for the other reasons discussed below, but at least we can forget about this...

**Comment by glyph** [comment]
@jothan Would you mind re-submitting just the type-annotation bits of this PR first, since they're not really related to the other stuff happening here, and would be useful to integrate on their own?

File: src/twisted/protocols/amp.py (hunk 14 above)
Comment (glyph) [reply src/twisted/protocols/amp.py:693]: If you have a key which is just a bit over 64k bytes, you end up screwing up the alternation of keys and values, which means that the failure mode involves the peer hanging forever; that's a pretty bad failure mode.

There's a way to do this which is *entirely* backwards compatible, though; just send the same key repeatedly, or send a key with `.1`, `.2`, etc appended to it.  This ends up adding 0.02% or so worth of overhead to the data (and if you have really large data, you want what's described on https://twistedmatrix.com/trac/ticket/2529 anyway, not just a way to circumvent the length-prefix limit).

You can also do this as a new parameter-type declaration, which also allows you to interrogate your peer's capabilities by checking for an `UNHANDLED` response if you have something you want to upgrade.

If we wanted to do something dramatically better, [netstrings](https://en.wikipedia.org/wiki/Netstring) are the way to go; arguably AMP should have used these in the first place.
